src/lib.rs

```diff
@@ -98,25 +98,177 @@ pub struct SelectionOutput {
+            let accumulated_weight: u32 = selected_coin
+                .iter()
+                .fold(0, |acc, &i| acc + inputs[i].weight);
+            let estimated_fee = 0;
+            let waste = calculate_waste(
+                inputs,
+                &selected_inputs,
+                &options,
+                accumulated_value,
+                accumulated_weight,
+                estimated_fee,
+            );
+            let selection_output = SelectionOutput {
+                selected_inputs: selected_coin,
+                waste: WasteMetric(waste),
+            };
+            Ok(selection_output)
+        }
+        None => Err(SelectionError::NoSolutionFound),
+    }
 }
 
-/// Return empty vec if no solutions are found.
+/// Return empty vec if no solutions are found
+///
+// changing the selected_inputs : &[usize] -> &mut Vec<usize>
 fn bnb(
     inputs_in_desc_value: &[(usize, OutputGroup)],
-    selected_inputs: &[usize],
-    effective_value: u64,
+    selected_inputs: &mut Vec<usize>,
+    acc_eff_value: u64,
     depth: usize,
-    bnp_tries: u32,
-    options: &CoinSelectionOpt,
-) -> Vec<usize> {
-    unimplemented!()
+    bnb_tries: &mut u32,
+    rng: &mut ThreadRng,
+    match_parameters: &MatchParameters,
+) -> Option<Vec<usize>> {
+    if acc_eff_value > match_parameters.target_for_match + match_parameters.match_range {
+        return None;
+    }
+    if acc_eff_value >= match_parameters.target_for_match {
+        return Some(selected_inputs.to_vec());
+    }
+
+    // Decrement of bnb_tries for every iteration
+    *bnb_tries -= 1;
+    // Capping the number of iterations on the computation
+    if *bnb_tries == 0 || depth >= inputs_in_desc_value.len() {
+        return None;
+    }
+    if rng.gen_bool(0.5) {
+        // exploring the inclusion branch
+        // first include then omit
+        let new_effective_value = acc_eff_value
+            + effective_value(
+                &inputs_in_desc_value[depth].1,
+                match_parameters.target_feerate,
+            );
+        selected_inputs.push(inputs_in_desc_value[depth].0);
+        let with_this = bnb(
+            inputs_in_desc_value,
+            selected_inputs,
+            new_effective_value,
+            depth + 1,
+            bnb_tries,
+            rng,
+            match_parameters,
+        );
+        match with_this {
+            Some(_) => with_this,
+            None => {
+                selected_inputs.pop(); // popping out the selected utxo if it does not fit
+                bnb(
+                    inputs_in_desc_value,
+                    selected_inputs,
+                    acc_eff_value,
+                    depth + 1,
+                    bnb_tries,
+                    rng,
+                    match_parameters,
+                )
+            }
+        }
+    } else {
+        /// Avoided creation of intermediate variable, for marginally better performance.
```

Comment:
remove this comment, confusing to the people who don't have context

Reply:
done